projects/frontend/src/draft-posts/index.md

```diff
@@ -0,0 +1,148 @@
+>
+> From [Wikipedia](https://en.wikipedia.org/wiki/Reverse_proxy)
+
+Essentially, a reverse proxy forwards traffic from a client to a set of servers behind the proxy.
+
+![Reverse Proxy Network](./assets/proxy.png)
+
+Why would you want to do this? There are many applications for reverse proxies. Load balancing, TLS termination, metric reporting, and A/B testing are just a few.
+
+If you’d like to learn more about proxying I highly recommend checking out  [Introduction to modern network load balancing and proxying](https://blog.envoyproxy.io/introduction-to-modern-network-load-balancing-and-proxying-a57f6ff80236)  by Matt Klein. Matt is the creator of [Envoy Proxy](https://www.envoyproxy.io/), a robust proxy server that powers service mesh tools like [Istio](https://istio.io/). His post does a great job of outlining the approaches used by modern load balancers and proxies.
+
+## Simple Go Reverse Proxy
+How do we go about rolling our own simple proxy in go?
+
+```go
+proxy := httputil.NewSingleHostReverseProxy(url)
+```
+
+Yep, with the go standard library it really is that simple.
+
+Let’s dig in here. The function returns a reverse proxy with the following method.
+
+```go
+func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+```
+
+All we need to do is configure the proxy and wire it up to a standard go http server to have a working reverse proxy as shown below.
+
+```go
+package main
+
+import (
+	"flag"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strconv"
+)
+
+func main() {
+	port := flag.Int("port", 8080, "port to listen on")
+	targetURL := flag.String("target-url", "", "downstream service url to proxy to")
+	flag.Parse()
+
+	u, err := url.Parse(*targetURL)
+	if err != nil {
+		log.Fatalf("Could not parse downstream url: %s", *targetURL)
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(u)
+
+	director := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		director(req)
+		req.Header.Set("X-Forwarded-Host", req.Header.Get("Host"))
+		req.Host = req.URL.Host
+	}
+
+	http.HandleFunc("/", proxy.ServeHTTP)
+	log.Printf("Listening on port %d", *port)
+	log.Fatal(http.ListenAndServe(":"+strconv.Itoa(*port), nil))
+}
+```
+
+That’s it! This server can now handle proxying HTTP 1 and 2 as well as web socket connections. You’ll notice that I’ve configured the _proxy.Director_ field. The _ReverseProxy.Director_ is a function that modifies the incoming request before it is forwarded. The signature is as follows:
+
+```go
+Director func(*http.Request)
+```
+
+A common use case for the director function is altering request headers. There is a default director set by  _httputil.NewSingleHostReverseProxy_ . The default implementation takes care of setting the request scheme, host, and path. I found this useful and didn’t want to duplicate the code. I had to reset the _req.Host_  field to contact HTTPS endpoints . I’ve also included an example of setting a request header.
```

Comment:
Given the go "Useful by default" mentality, I might say something like this instead:
```suggestion
A common use case for the director function is modifying request headers. The default director implementation for _httputil.NewSingleHostReverseProxy_ takes care of setting the request Scheme, Gost, and Path. I found this useful and didn’t want to duplicate the code, so I wrapped that implementation. *Note*: I had to reset the _req.Host_  field to handle HTTPS endpoints. I’ve also included an example of setting a request header.
```

Reply:
Two questions:
1. What are the implications of overwriting the `req.Host` property?
2. Does `req.Header.Set` overwrite or append? This is important as I found out at work a couple months ago and I would mention it.